src/sinks/s3_common/sink.rs

```diff
@@ -78,16 +231,77 @@ where
     }
 }
 
+struct PayloadBuilder {
+    framer: Framer,
+    compression: Compression,
+    encoder: crate::codecs::Encoder<Framer>,
+}
+
+struct Payload {
+    body: Bytes,
+    uncompressed_size: usize,
+}
+
+impl PayloadBuilder {
+    fn new(serializer: Serializer, framer: Framer, compression: Compression) -> Self {
+        // We don't actually use the serializer at all, we just need an `Encoder` for the batch
+        // prefix and suffix, which for some reason you can't get from the framer alone.
+        let encoder = crate::codecs::Encoder::<Framer>::new(framer.clone(), serializer);
+        Self {
+            framer,
+            compression,
+            encoder,
+        }
+    }
+
+    fn build_payload(&self, encoded_events: Vec<BytesMut>) -> crate::Result<Payload> {
+        let mut compressor = Compressor::from(self.compression);
+
+        let mut remaining = encoded_events.len();
+        let mut bytes_written = 0;
+
+        let batch_prefix = self.encoder.batch_prefix();
+        write_all(&mut compressor, remaining, batch_prefix)?;
+        bytes_written += batch_prefix.len();
+        for mut event in encoded_events {
+            // TODO: Not framing the final value matches existing behavior but is almost certainly a
+            // bug for some types of framing (e.g. length-delimited) and questionable for others
+            // (e.g. newline- and character-delimited). Really it only seems appropriate for
+            // something like JSON arrays where trailing commas are disallowed.
+            if remaining > 0 {
+                // This is confusing, but how our codecs work currently. We pass the serialized
+                // payload as the destination BytesMut, which the framer then splits and rewrites
+                // with framing (really just the framing delimiter; we have to handle batch prefix
+                // and suffix ourselves) into the original.
+                self.framer.clone().encode((), &mut event)?
+            }
```

Comment:
This seems to strongly point towards needing different handling in the traits, with separate handling for encoding the ultimate event. That is, of course, out of scope for this PR except in so far as this exposes a bug there. Will this cause an issue if the object is extended, for example when we get around to the large object handling improvements?